### ISSUE TITLE
3.8+ argparse allow_abbrev=False is broken

### ISSUE THREAD

**Issue by Fox214** [python/cpython#94756]
Sample code:
```
import argparse
import os
import sys

# Parse command line options
usage='%(prog)s [options] file1'
        
cli = argparse.ArgumentParser(usage=usage,add_help=False,allow_abbrev=False)
cli.add_argument("file1", nargs='?', help="somefile") 
cli.add_argument('-h', '--help',
               action='store_true', dest='print_help',
               help='Print this menu and exit.')
cli.add_argument('-f', '--fun',
                action='store', dest='fun', default=0, metavar='STRING',
                help='prints string.')
cli.add_argument('-a', 
                action='store_true', dest='flag',
                help='an option')
cli.add_argument('--func',
                action='store_true', dest='func1',
                help='func help')
cli.add_argument('--funcHelp',
                action='store_true', dest='func_help',
                help='FUNC HELP')


called=""
for a in sys.argv:
    called+=a + " "
    
print("Called as: ", called)
print("version",sys.version)
try:
    args = cli.parse_args() 
except SystemExit:
    print("Invalid options used, please check the help menu")
    sys.exit(33)

if args.print_help:
    cli.print_help()
    sys.exit(0)

if args.fun:
    print(f"fun set {args.fun}")
   
if args.func_help:
    print("func_help set")
    
if args.func1:
    print("func1 set")
    
if args.flag:
    print("an option set")

if args.file1:
    print(f"file1: {args.file1}")

if not args.fun and not args.func1 and not args.func_help and not args.flag:
    if not args.file1:
        cli.print_usage()
        print("Must provide an argument")
        sys.exit(33)


sys.exit(0)
```

Result in 3.8+ (incorrect):
```
Called as:  ./TestAllowAbbrev.py -fa 
version 3.8.5 (default, Jul 11 2022, 05:20:48) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
fun set a
```

(a is supposed to be its own option, not an input to -f) 

Result in 3.7- (correct):
```
Called as:  ./TestAllowAbbrev.py -fa 
version 3.7.4 (default, Jan 28 2021, 06:22:35) 
[GCC 6.2.0]
usage: TestAllowAbbrev.py [options] file1
TestAllowAbbrev.py: error: unrecognized arguments: -fa
Invalid options used, please check the help menu
```

using a multicharacter option with an argument with a single dash is also broken if the argument with an input is a subset of some other option as it uses the rest of the word as an input. (thus clearly allowing abbreviations when explicitly told not to)

In argparse.py for python 3.8.4 moving if self.allow_abbrev: from line 2187 to line 2147 fixed this issue for me.

**Comment by yourlefthandman** [comment]
Related to https://bugs.python.org/issue39546

**Comment by lucasalavapena** [comment]
I tried to reproduce the issue, and my first impression is that this is the expected behavior. Afterall, both `-f` and `--func` are arguments here so `allow_abbrev=False` is irrelevant if you are calling the full argument as stated by `./TestAllowAbbrev.py -fa` which be interpreted as `-f a`

Also note I see the same behaviour as shown for this script in 3.8, in 3.10.9 and 3.12.0a5.

**Comment by Fox214** [comment]
The problem is if -a -f is supposed to be synonymous with -af then -fa and -af should behave equivalently, they do not, once produces and expected error the other does not. The issue is that its prioritizing the abbreviated option over the parsing of the entire word. 

Both options (-af and -fa) fail (expectedly) in 3.7 and behave differently in 3.8 if its prioritizing daisy chaining over allowing abbreviations (thus perhaps the illusion that's its allowing abbreviations, because a user is more likely to interpret the input word over a daisy chain, especially as the amount of input options grows) 

I would probably rather turn off the ability to chain options with a single - than deal with this sort of behavior. The order should not matter.

**Comment by lucasalavapena** [comment]
@Fox214 If `-f`(`--fun`) has `action='store_true'` (with other now irrelevant args removed) then the behaviour is as you expect, with  `action='store'` (which implies you want to pass in a value after the argument), I do not see a reason to complain about the behaviour unless you want to explicitly require a space between the argument and the argument's value like in 3.7 (see below).


```
(py37) [XXX cpython]$ python gh-83727.py -fa
Called as:  gh-83727.py -fa
version 3.7.16 (default, Jan 17 2023, 22:20:44)
[GCC 11.2.0]
usage: gh-83727.py [options] file1
gh-83727.py: error: unrecognized arguments: -fa
Invalid options used, please check the help menu
(py37) [XXX cpython]$ python gh-83727.py -f a
Called as:  gh-83727.py -f a
version 3.7.16 (default, Jan 17 2023, 22:20:44)
[GCC 11.2.0]
fun set a
```
For completeness here it is in 3.12:
```
[XXX cpython]$ ./python gh-83727.py -fa
Called as:  gh-83727.py -fa 
version 3.12.0a5+ (heads/main:7346a381be, Feb 21 2023, 14:41:15) [GCC 12.2.1 20230201]
fun set a
[XXX cpython]$ ./python gh-83727.py -f a
Called as:  gh-83727.py -f a 
version 3.12.0a5+ (heads/main:7346a381be, Feb 21 2023, 14:41:15) [GCC 12.2.1 20230201]
fun set a
```

**Comment by Fox214** [comment]
Yes the 3.7 behavior is desirable. Calling options in different orders should not affect the output. I have been manually editing the argparse.py file myself to fix (its simple) but remember to do so is not desirable.  (try it with -fa and -af)

**Comment by hpaulj** [comment]
In 3.10, with this parser (just short options, so we don't need to worry about abbreviations):

    p1 = argparse.ArgumentParser()
    p1.add_argument('-f')
    p1.add_argument('-a',action='store_true');

This looks good to me, parsing `-f` as the 'store' with a following argument:

    In [43]: p1.parse_args(['-fa'])
    Out[43]: Namespace(f='a', a=False)

same as:  `p1.parse_args(['-f', 'a'])`  and `p1.parse_args(['-f=a'])`.  These may be less ambiguous, but I don't see why '-fa' should raise an error.

This error makes sense:

     In [46]: p1.parse_args(['-af'])
     usage: ipykernel_launcher.py [-h] [-f F] [-a]
     ipykernel_launcher.py: error: argument -f: expected one argument

Correct usage:

    In [47]: p1.parse_args(['-af','1'])      # also with '-af1'
    Out[47]: Namespace(f='1', a=True)

chaining and '=' don't work together, but that's to be expected.  Chaining is a short hand when working with short flags; '=' is intended for long flags':

    In [49]: p1.parse_args(['-af=1'])
    Out[49]: Namespace(f='=1', a=True)

I can't test the earlier pre-"allow-abbreviation' behavior.

While in general, the order of flagged optionals shouldn't matter, but in the context of chaining order does matter.  All but the last flag must be of the 'nargs=0' type.  Only the last is allowed to take arguments.

Conventionally, short flags take one dash, long ones two.  `argparse` doesn't exactly enforce this.  Double dash are given preference when setting the default `dest`.  And only single character single dash flags work with chaining.  Otherwise I don't think it pays much attention to the distinction between short and long.

**Comment by Fox214** [comment]
The bottom line is the 3.7- behavior is desirable, the 3.8+ behavior is not. The ability to specify the preference order (word, chain, nargs) or disable chaining of nargs>0  or disable chaining entirely as additional options may also fix the problem (my real case is much more complex and does not make use of -- at all). Traditionally flags were single -, the gnu -- standard is much newer (and their own compiler doesn't follow it). migrating 'old' things (or families or utilities related to said things) it is not desirable to change the usage format (or possible based on style guidelines)

**Comment by serhiy-storchaka** [comment]
This is an expected behavior. Many argument parsers in Python and other programming languages behave the same. It goes far to early versions of Unix, and may even predate it. `-f` requires an argument, so `-fa` means `-f` with argument `a`, and `-af` is an error.

**Comment by Fox214** [comment]
Apologies if I sound rude here but this is really frustrating because I feel like I am living in a different reality. 

I cannot vouch if "Many argument parsers in Python and other programming languages behave the same" but in my 30 years of using unix, I have never encountered this, so this statement seems to be false. I decided intentionally look for this and the very first program I tried "find" (a common unix util) simply disallows abbreviations and rejects the wrong argument "find . -g <some group name>" yields: find: unknown predicate `-g', **I could not have python argparse work like find in 3.8 or newer python.** 

Guessing is bad, and argparse is guessing wrong. I want to disable the guessing, why is this request deemed unreasonable? I've even provided a solution, and while I would argue the default behavior is bad, provided I have a way to actually disable it, I don't really care if I need to update the code. 

The argument about precedence of chaining vs. abbreviations is honestly a side discussion, I do not want to allow abbreviations (this just underscores the reason, as the user may not get their intent because they think they're doing one but the program does the other- so I would rather error out immediately and they can fix their arguments).

FWIW, finding another program that behaves the same way is not bolstering the argument because the behavior I want is none of the above. Don't break the user space or something....

**Comment by serhiy-storchaka** [comment]
Try `ls -aw 40` and `ls -wa 40`. This is similar to the original example.

`find` is an outlier among Unix utilities. It supports single-dash long options and does not support short option grouping (it has not many short options anyway). It does not support long option abbreviation.

In `argparse`, you can disallow long option abbreviation by passing `allow_abbrev=False`. But it was broken for single-dash long options. Now it is fixed. Wait for the next bugfix release of Python 3.12.

Short option grouping cannot be disallowed. You can open a new issue for adding a new option to disallow it. It can be implemented in one of future Python versions. Since it is a new feature, not a bug, it may have lesser priority. But if you are lucky, it may be implemented in hours.

I am sorry that fixing this bug taken so much time. We have too many issues, too little volunteers, and there was no active maintainer of the `argparse` module.

**Comment by Fox214** [comment]
> In `argparse`, you can disallow long option abbreviation by passing `allow_abbrev=False`. But it was broken for single-dash long options. Now it is fixed. Wait for the next bugfix release of Python 3.12.
> 

I will test this out in 3.12.7? I only have single '-' and that may fix my complaint.

**Comment by serhiy-storchaka** [comment]
Actually, no. Because your example does not have single-dash long options, so that change will not affect it. You complain about short options grouping. Currently you cannot disallow it. `allow_abbrev=False` does not affect it. It was time when `allow_abbrev=False` disallowed also short options grouping, but it was a bug, and it was fixed in 3.8 or like.

**Comment by Fox214** [comment]
The original problem was with '-' I put double in the example to avoid discussion about '-' vs '--' for long options, I think I had found another complaint (similar to mine, and the discussion ended with '-' vs '--' instead of focusing on the fact the '-' was broken and forced abbreviations, (I don't recall the specifics nor do I feel like finding the link).
